src/main/java/xyz/nucleoid/plasmid/util/BucketScanner.java

```diff
@@ -0,0 +1,231 @@
+     */
+    public static LongSet findSix(BlockPos origin, int depth, Predicate<BlockPos> predicate) {
+        LongSet set = new LongArraySet();
+        Deque<BlockPos> ends = new ArrayDeque<>();
+        ends.push(origin);
+        BlockPos.Mutable local = origin.mutableCopy();
+        while(depth > 0) {
+            if(ends.isEmpty()) {
+                return set;
+            }
+            BlockPos pos = ends.pollLast();
+            for(Direction direction : Direction.values()) {
+                local.set(pos.offset(direction));
+                if(predicate.test(local) && !set.contains(local.asLong()) && !ends.contains(local)) {
+                    ends.push(local.toImmutable());
+                }
+            }
+            set.add(pos.asLong());
+            depth--;
+        }
+        return set;
+    }
+
+    /**
+     * Finds any 18-connected blocks and puts them in a {@link LongSet}.
+     *
+     * @param origin   the position of the first block
+     * @param depth    the amount of maximum blocks to find
+     * @param ruleTest the rule test for the blocks that can be accepted in the set
+     */
+    public static LongSet findEighteen(BlockPos origin, int depth, RuleTest ruleTest, ServerWorld world, Random random) {
+        return findEighteen(origin, depth, pos -> ruleTest.test(world.getBlockState(pos), random));
+    }
+
+    /**
+     * Finds any 18-connected blocks and puts them in a {@link LongSet}.
+     *
+     * @param origin the position of the first block
+     * @param depth  the amount of maximum blocks to find
+     * @param tag    the tag for the blocks that can be accepted in the set
+     */
+    public static LongSet findEighteen(BlockPos origin, int depth, Tag<Block> tag, ServerWorld world) {
+        return findEighteen(origin, depth, pos -> world.getBlockState(pos).isIn(tag));
+    }
+
+    /**
+     * Finds any 18-connected blocks and puts them in a {@link LongSet}.
+     *
+     * @param origin the position of the first block
+     * @param depth  the amount of maximum blocks to find
+     * @param block  the block type that can be accepted in the set
+     */
+    public static LongSet findEighteen(BlockPos origin, int depth, Block block, ServerWorld world) {
+        return findEighteen(origin, depth, pos -> world.getBlockState(pos).isOf(block));
+    }
+
+    /**
+     * Finds any 18-connected blocks and puts them in a {@link LongSet}..
+     *
+     * @param origin    the position of the first block
+     * @param depth     the amount of maximum blocks to find
+     * @param predicate the predicate for the blocks that can be accepted in the set
+     */
+    public static LongSet findEighteen(BlockPos origin, int depth, Predicate<BlockPos> predicate) {
+        LongSet set = new LongArraySet();
+        Deque<BlockPos> ends = new ArrayDeque<>();
+        ends.push(origin);
+        BlockPos.Mutable local = origin.mutableCopy();
+        while(depth > 0) {
```

Comment:
It seems like this isn't depth but rather just a count- I imagine this should be depth in the sense of how many steps have been taken from the source block?

Reply:
Yeah I guess count would make more sense here

Reply:
Count would be more accurate to the existing implementation but depth is more useful

Reply:
Hmm.. what about `amount`? Also, why would depth be more useful?

Reply:
It’s hard to relate a count to what the method will scan- especially since it depends heavily on the implementation. If it’s breadth-first or depth-first, or what order it iterates neighbours, very different results will be selected.

Reply:
Oh, alright, well I could implement both ideas with a simple boolean

Reply:
Ugh, actually I think I'll stick with depth-first

Reply:
Added both methods in e6cfa44 😄